lib/configSchema/index.test.js

```diff
@@ -0,0 +1,234 @@
+        config.plugins = ['first', 'second'];
+      },
+    },
+
+    {
+      isValid: true,
+      description: 'resources free form object',
+      mutation: () => {
+        config.resources = {
+          Resources: {
+            SomeBucket: {
+              some: 'prop',
+            },
+          },
+        };
+      },
+    },
+
+    {
+      isValid: true,
+      description: 'functions',
+      mutation: () => {
+        config.functions = {
+          someFunc: {
+            handler: 'handler.main',
+            events: [],
+            name: 'some-service-dev-someFunc',
+          },
+        };
+      },
+    },
+    {
+      isValid: true,
+      description: 'function with parameters that do not exist in schema',
+      mutation: () => {
+        config.functions.someFunc = {};
+        config.functions.someFunc.handler = 'some.handler';
+        config.functions.someFunc.memorySize = 512;
+        config.functions.someFunc.reservedConcurrency = 5;
+        config.functions.someFunc.provisionedConcurrency = 3;
+        config.functions.someFunc.runtime = 'nodejs12.x';
+      },
+    },
+
+    {
+      isValid: true,
+      description: 'provider name',
+      mutation: () => {
+        config.provider.name = 'aws';
+      },
+    },
+    {
+      isValid: false,
+      description: 'provider no name',
+      mutation: () => {
+        delete config.provider.name;
+      },
+    },
+    {
+      isValid: false,
+      description: 'provider not existing name',
+      mutation: () => {
+        config.provider.name = 'azureeee';
+      },
+    },
+
+    {
+      isValid: true,
+      description: 'package',
+      mutation: () => {
+        config.package = {
+          individually: true,
+          path: undefined,
+          artifact: 'path/to/my-artifact.zip',
+          exclude: ['.git/**', '.travis.yml'],
+          include: ['src/**', 'handler.js'],
+          excludeDevDependencies: false,
+        };
+      },
+    },
+    {
+      isValid: false,
+      description: 'package',
+      mutation: () => {
+        config.package = {
+          unknownProp: 'shoud be invalid',
+        };
+      },
+    },
+
+    {
+      isValid: true,
+      description: 'layers free form object',
+      mutation: () => {
+        config.layers = { hello: { path: 'some-path' } };
+      },
+    },
+    {
+      isValid: false,
+      description: 'layers defined as string',
+      mutation: () => {
+        config.layers = 'some string';
+      },
+    },
+
+    {
+      isValid: false,
+      description: 'outputs',
+      mutation: () => {
+        config.outputs = {
+          some: {
+            outputs: 'they are deprecated',
+          },
+        };
+      },
+    },
+  ];
+
+  for (const someCase of cases) {
+    const message = someCase.isValid ? 'pass' : 'fail';
+    // eslint-disable-next-line no-loop-func
+    it(`should ${message} validation for ${someCase.description}`, () => {
+      someCase.mutation();
```

Comment:
Concerning tests, it'll be nicer to rely on `runServerless` util, as only through initalization of complete serverless instances we can reliably confirm that validation works as expected.

(at this point all new tests are written with `runServerless`)

Some guidelines, so you don't feel stuck:

Documentation of `runServerless` util is here: https://github.com/serverless/test/blob/48e348002378071319f9e0e58e127b4863f8851e/docs/run-serverless.md

In context of this repository we should require [tests/utils/run-serverless.js](https://github.com/serverless/serverless/blob/73107822945a878abbdebe2309e8e9d87cc2858a/tests/utils/run-serverless.js) (it presets path to 
`serverless` framework).

Examples of tests:
- Run against pre-prepared fixture: https://github.com/serverless/serverless/blob/73107822945a878abbdebe2309e8e9d87cc2858a/lib/plugins/aws/package/compile/functions/index.test.js#L2629-L2632
- Run against config passed inline: https://github.com/serverless/serverless/blob/73107822945a878abbdebe2309e8e9d87cc2858a/lib/plugins/aws/package/lib/generateCoreTemplate.test.js#L11-L14

I believe in this case, we need to create a fixture (to be placed in [tests/fixtures](https://github.com/serverless/serverless/tree/73107822945a878abbdebe2309e8e9d87cc2858a/tests/fixtures)) with some dummy provider plugin included (as we need to have `provider.name`, and just this file should not define options for it).

For fixture I imagine base `serverless.yml` as:

```yaml
service: test
provider:
  name: test
plugins:
  - ./test-provider
```

and locally put `test-provider.js` could be as:

```javascript
class TestProviderPlugin {
  constructor(serverless) {
    // extend schema to support 'test' provider (just name)
  }
}
```

Note: You may use above fixture to test all config variants you want to test. As in tests you can easily create modified copies of it via `fixtures.extend` as e.g. it was done here: https://github.com/serverless/serverless/blob/73107822945a878abbdebe2309e8e9d87cc2858a/lib/plugins/aws/package/compile/functions/index.test.js#L2647-L2654

Probably the most appropriate command now to use in testing is `sls info`

If you have any doubts, questions, please ask. I don't want this to feel as blocker

Reply:
Thanks for pointing to this `runServerless` util. I was able to refactor this file to use it, with the help of `modulesCacheStub` option that helps me to keep track of console output.

I was, however, unable to use fixtures extend for `plugins`, `package` and `layers` sections. So I removed those tests for now.

I am planning to add expected error message properly to cases that are expected to fail. This way the test would ensure that the correct error message is shown.

The tests became slower but I think we can live with that. The benefit is that it integratively ensures that validation works as expected.

One more tiny thing. I'm getting `(node:81439)  MaxListenersExceededWarning` warning if 10+ tests length is set:

```
  #configSchema
    ✓ should pass validation for config (1441ms)
    ✓ should fail validation for service name (883ms)
    ✓ should fail validation for service awsKmsKeyArn (892ms)
    ✓ should pass validation for service awsKmsKeyArn (847ms)
    ✓ should fail validation for service unknown property (831ms)
    ✓ should pass validation for app name (817ms)
    ✓ should pass validation for org name (827ms)
    ✓ should pass validation for custom object (841ms)
    ✓ should pass validation for resources free form object (940ms)
(node:81439) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 exit listeners added to [process]. Use emitter.setMaxListeners() to increase limit
    ✓ should pass validation for functions (820ms)
    ✓ should pass validation for function with parameters that do not exist in schema (849ms)
    ✓ should fail validation for provider not existing name (818ms)
    ✓ should pass validation for package (868ms)


  13 passing (12s)
```

I'm not sure where I should put `setMaxListeners(20)`. Or should we just ignore this for now?

Reply:
> The tests became slower but I think we can live with that. The benefit is that it integratively ensures that validation works as expected.

Yes, exactly

> I'm not sure where I should put setMaxListeners(20). Or should we just ignore this for now?

Let's ignore it. It happens also for some tests we have in master. As investigated it once, it's harmless and hard to workaround with current framework design. IMO definitely not a concern to address in scope of this PR.